### ISSUE TITLE
CLI: Refactor to provide only top level commands

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#8925]
### Use case description

Currently `serverless` CLI exposes some sub level commands as:

```
sls deploy function
sls invoke local
```

It'll make internal handling easier, if commands are only top-level as:

```
sls deploy-function
sls invoke-local
```

Reasoning:
1. Straightforward to translate programmatic API, where it's straightfoward to have `deploy()` and `deployFunction()`, and not `deploy()` and `deploy.function()`
2. Non-ambiguous CLI args parsing. With multi levels  it's not clear whether in `sls deploy function foo` command is `deploy function` or `deploy function foo`. While allowing just top level, we can accepts positional arguments, so we can have `sls deploy-function foo` instead of `sls deploy function -f foo` as we have now.
3. All popular CLI's I'm aware of (as `git`, `npm`) do not support sub command levels

### Proposed solution

_TBA_

--- 

@pgrzesik what do you think?

**Comment by pgrzesik** [comment]
Thanks @medikoo for that proposal. I agree with the reasoning, especially with the number 2, however, there are a few things that I'm potentially worried about:
- Documentation upgrade + all existing materials will be invalid if they depend on commands that will be deprecated - I guess there's nothing that we can really do about it 
- Plugins - would that require plugins to adapt to that as well? Do we even support subcommands for plugins?

**Comment by medikoo** [comment]
> Documentation upgrade + all existing materials will be invalid if they depend on commands that will be deprecated - I guess there's nothing that we can really do about it

Yes, I think we can relatively easily upgrade documentation as is in this repository, but ofc any blog articles will remain outdated.

I think it's inevitable if we want to really move forward with the Framework, and not remain fixed on what it is currently.

> Plugins - would that require plugins to adapt to that as well? Do we even support subcommands for plugins?

Some plugins expose their own commands, but I'm not sure if any are sub nested. Anyway I expect changes to how commands are setup in general, so anyway plugins will have to be updated at some point.

And when that is in picture I believe that we should also actively help drive their migration to new changes, as with time they will get more breaking, and plugins are really important in our ecosystem.

I propose that we:
1. Prepare a solid plugin authors documentation and template at https://github.com/serverless/serverless/tree/c8547e87520ad0f7dbfb3e083a4399ba43e49703/lib/plugins/create/templates/plugin
2. Pick ca 50 most popular and active plugins and adapt them to meet the quality constraints:
   1. Be versioned against v2 `serverless` major release (ensure Framework is listed in `peerDependencies`, and maybe add some extra programmatic validation)
   2. Ensure they safe to run in v2 (e.g. do not depend on Bluebird promises)
   3. Ensure they do not trigger deprecations (e.g. have schema for properties they add on defined)
   4. With every new major release reiterate on them, and ideally make them ready for a new version, at the time of release
   
I believe we should raise this matter on next meeting with @ac360

**Comment by pgrzesik** [comment]
Thanks for feedback @medikoo :bow: As for plugins - I totally agree with your proposal as I believe plugins are really important for a lot of Framework users and ultimately they are responsible for better adoption of the Framework. I agree with raising that issue on our next meeting with @ac360 - in the meantime I'll try to update the documentation, I've also updated the issues in plugins that I've mentioned yesterday :+1: 

Btw - I've also noticed that some plugins in fact use `command subcommand` approach.
